packages/workspace/src/browser/workspace-commands.ts

```diff
@@ -308,18 +311,26 @@ export class WorkspaceCommandContribution implements CommandContribution {
      *
      * @param name the simple file name of the file to validate.
      * @param parent the parent directory's file stat.
+     * @param recursive allow file or folder creation using recursive path
+     * @param dialog optional SingleTextInputDialog instance
      */
-    protected validateFileName(name: string, parent: FileStat): string {
-        if (!validFilename(name)) {
-            return 'Invalid name, try other';
+    protected validateFileName(name: string, parent: FileStat, recursive: boolean = false, dialog?: SingleTextInputDialog): string {
+        if (!name) {
+            return '';
         }
-        if (parent.children) {
-            for (const child of parent.children) {
-                if (new URI(child.uri).path.base === name) {
-                    return 'A file with this name already exists.';
-                }
-            }
+        if (!recursive && !validFilename(name)) {
+            return 'Invalid file or folder name';
         }
+        const fileName = path.basename(name);
+        if (!validFilename(fileName) || name.startsWith('/') || name.startsWith(' ') || name.endsWith(' ')) {
+            return `The name <strong>${fileName}</strong> is not a valid file or folder name.`;
+        }
+        const childUri = new URI(parent.uri).resolve(name).toString();
+        this.fileSystem.exists(childUri).then(exists => {
+            if (exists && dialog) {
+                dialog.error(`A file or folder <strong>${fileName}</strong> already exists at this location.`);
```

Comment:
We should change a type of validation function to `MaybPromise<DialogError>` to allow async validation. Inside the dialog we will need to make sure that concurrent validation does not occur, i.e. queue them or cancel.